File: pixijs/game/bullet.js
export class Bullet {
  constructor() {
    this.reset();
  }

  reset() {
    this.x = 0;
    this.y = 0;
    this.vx = 0;
    this.vy = 0;
    this._baseVx = 0;
    this._baseVy = 0;
    this.speed = 0;
    this.damage = 0;
    this.owner = 'player'; // 'player' or 'enemy'
    this.penetrate = 0;
    this._basePenetrate = 0;
    this.ricochet = false;
    this._baseRicochet = false;
    this.color = 0xffffff;
    this.maxRange = 1000;
    this.distanceTraveled = 0;
    this.hitCount = 0;
    this.isCrit = false;
    this.aimCritTarget = null;
    this.aimCritMult = 1;
    this.enhancedPierceActive = false;
    this.hitEntities = new Set(); // To avoid hitting same entity multiple times (if piercing)
    this._lastRoomBonus = null;
    this._ricocheted = false;
    this.isDead = false;
    this._trailTimer = 0;
    this._rangeDecayMult = 1;
    
    // Future modifiers support
    this.modifiers = [];
  }

  init(data) {
    this.x = data.x;
    this.y = data.y;
    this.vx = data.vx;
    this.vy = data.vy;
    this._baseVx = data.vx;
    this._baseVy = data.vy;
    this.speed = Math.hypot(data.vx, data.vy);
    this.damage = data.damage || 1;
    this.owner = data.owner || 'player';
    this.penetrate = data.penetrate ?? 0;
    this._basePenetrate = this.penetrate;
    this.ricochet = !!data.ricochet;
    this._baseRicochet = this.ricochet;
    this.color = data.color || 0xffffff;
    this.maxRange = data.maxRange || 1000;
    this.isCrit = !!data.isCrit;
    this.aimCritTarget = data.aimCritTarget || null;
    this.aimCritMult = data.aimCritMult || 1;
    this.isDead = false;
    
    this.distanceTraveled = 0;
    this.hitCount = 0;
    this.enhancedPierceActive = false;
    this.hitEntities.clear();
    this._lastRoomBonus = null;
    this._ricocheted = false;
    this._rangeDecayMult = 1;
  }
}

const pool = [];

export function acquireBullet(data) {
  const b = pool.pop() ?? new Bullet();
  b.init(data);
  return b;
}

export function releaseBullet(b) {
  b.isDead = true;
  pool.push(b);
}

File: Scripts/Combat/Bullet.cs
using System.Collections.Generic;
using Godot;

namespace SpaceOrLife.Combat;

public sealed class Bullet
{
    public Vector2 Position;
    public Vector2 Velocity;
    public Vector2 BaseVelocity;
    public float Speed;

    public BulletOwner Owner = BulletOwner.Player;
    public float Damage;
    public int Penetrate;
    public int BasePenetrate;
    public bool Ricochet;
    public bool BaseRicochet;

    public Color Color = Colors.White;
    public float MaxRange = 1000f;
    public float DistanceTraveled;
    public int HitCount;

    public bool IsCrit;
    public bool IsIncendiary;
    public bool IsFreeze;
    public float HitStun;
    public IDamageable? AimCritTarget;
    public float AimCritMult = 1f;

    public bool IsDead;
    public float TrailTimer;
    public float RangeDecayMult = 1f;
    public string? LastRoomBonus;
    public bool Ricocheted;
    public float LastSpeedMult = 1f;

    public readonly HashSet<IDamageable> HitEntities = new();

    public void Init(BulletSpawnData data)
    {
        Position = data.Position;
        Velocity = data.Velocity;
        BaseVelocity = data.Velocity;
        Speed = data.Velocity.Length();

        Owner = data.Owner;
        Damage = data.Damage;
        Penetrate = data.Penetrate;
        BasePenetrate = data.Penetrate;
        Ricochet = data.Ricochet;
        BaseRicochet = data.Ricochet;

        Color = data.Color;
        MaxRange = data.MaxRange;

        IsCrit = data.IsCrit;
        IsIncendiary = data.IsIncendiary;
        IsFreeze = data.IsFreeze;
        HitStun = data.HitStun;
        AimCritTarget = data.AimCritTarget;
        AimCritMult = data.AimCritMult;

        IsDead = false;
        DistanceTraveled = 0f;
        HitCount = 0;
        TrailTimer = 0f;
        RangeDecayMult = 1f;
        LastRoomBonus = null;
        Ricocheted = false;
        LastSpeedMult = 1f;
        HitEntities.Clear();
    }

    public void Reset()
    {
        Position = Vector2.Zero;
        Velocity = Vector2.Zero;
        BaseVelocity = Vector2.Zero;
        Speed = 0f;

        Owner = BulletOwner.Player;
        Damage = 0f;
        Penetrate = 0;
        BasePenetrate = 0;
        Ricochet = false;
        BaseRicochet = false;

        Color = Colors.White;
        MaxRange = 1000f;
        DistanceTraveled = 0f;
        HitCount = 0;

        IsCrit = false;
        IsIncendiary = false;
        IsFreeze = false;
        HitStun = 0f;
        AimCritTarget = null;
        AimCritMult = 1f;

        IsDead = true;
        TrailTimer = 0f;
        RangeDecayMult = 1f;
        LastRoomBonus = null;
        Ricocheted = false;
        LastSpeedMult = 1f;
        HitEntities.Clear();
    }

    public void SyncBaseVelocity()
    {
        if (LastRoomBonus == "speeddown" || IsWindBonus(LastRoomBonus))
        {
            var mult = LastSpeedMult != 0f ? LastSpeedMult : 1f;
            BaseVelocity = Velocity / mult;
        }
        else
        {
            BaseVelocity = Velocity;
        }
    }

    private static bool IsWindBonus(string? bonus)
    {
        return bonus == "wind_east" || bonus == "wind_west" ||
               bonus == "wind_north" || bonus == "wind_south";
    }
}

public sealed class BulletSpawnData
{
    public Vector2 Position;
    public Vector2 Velocity;
    public BulletOwner Owner = BulletOwner.Player;
    public float Damage = 1f;
    public int Penetrate;
    public bool Ricochet;
    public Color Color = Colors.White;
    public float MaxRange = 1000f;
    public bool IsCrit;
    public bool IsIncendiary;
    public bool IsFreeze;
    public float HitStun;
    public IDamageable? AimCritTarget;
    public float AimCritMult = 1f;
}
